radial_build_tools.py

```diff
@@ -56,6 +54,36 @@ def __init__(self, build, **kwargs):
         ):
             self.__setattr__(name, kwargs[name])
 
+        self.used_colors = set()
+        self.available_colors = set(matplotlib.colors.XKCD_COLORS.values())
+        self.colors = self.assign_colors()
+
+    def assign_colors(self):
+        """
+        Assign colors to each layer. If a color is provided, use it.
+        If not, generate a random color that hasn't been used yet.
+        """
+        colors = []
+        for layer in self.build.values():
+            if "color" not in layer:
+                layer["color"] = random.choice(list(self.available_colors))
+            color = layer["color"]
+            self.used_colors.add(color)
+            self.available_colors.discard(color)
+
+            colors.append(color)
+
+        return colors
+
+    def generate_unique_color(self):
+        """
+        Generate a random color that has not been used yet.
+        """
+        color = random.choice(list(self.available_colors))
+        self.available_colors.remove(color)  # Remove chosen color
+        self.used_colors.add(color)  # Add to used colors
+        return color
```

Comment:
```suggestion
        self.color_pool = set(mcolors.CSS4_COLORS) | set(mcolors.XKCD_COLORS)
        self.colors = self.assign_colors()
    
    def assign_colors(self):
        """
        Assigns unique colors to each layer in the build, using a combination of
        CSS4 and XKCD colors.
        """
        layer_colors = []
        used_colors = set()
        
        # First pass handles pre-specified colors
        for layer in self.build.values():
            if color := layer.get('color'):
                used_colors.add(color)
                layer_colors.append(color)
            else:
                layer_colors.append(None)
        
        # Create pool of available colors by removing used ones, eliminating the possibility of randomly selecting a specified layer color
        available_colors = list(self.color_pool - used_colors)
        
        # Second pass assign random colors to the unspecified layers
        for i, color in enumerate(layer_colors):
            if color is None:
                new_color = random.choice(available_colors)
                available_colors.remove(new_color)
                layer_colors[i] = new_color
        
        return layer_colors
```
My reasoning for this method is that the earlier one leaves open the possibility for a color to be randomly selected before a layer with that assigned color is called (which would then result in it being added to the used_colors list, stopping it from being used later) this would result in two layers having the same color unintentionally. I increased the color library so it's easier for users to assign a color by name instead of hexadecimal. I'll move it out of the __init__ later.

Reply:
That makes sense - thanks for the clarification... I wonder if there's still a way to streamline it somewhat.

Reply:
Yeah, I'm not really a fan of using two loops. I'll play around with it this week and see if I find something better.

Reply:
I think it may always take 2 loops - but you could grab all the assigned colors with a list comprehension and then remove them all from the available colors, and then loop through everything again?

It might be best to leave it how you originally had it, with a comment about why you have two loops.